Tensor Prototype

dabrahams opened the pull request:
Hi Ed,

This branch tracks my work on the tensor prototype you asked for.  Lots more to come…

ewconnell commented:
In Math.swift I'm still getting `1. Cannot differentiate functions that have not been marked '@differentiable' and that are defined in other files`

Commit:
Initial prototyping starts with my gist:

https://gist.github.com/dabrahams/1fa37dd518cac97209db5fedbfc09ce0

Commit:
Break out StaticArray definition

Commit:
Use XCTest assertions in our test

Commit:
Renaming

StaticArray -> FixedSizeArray
test_StaticArray.swift -> test_Prototype.swift

Commit:
Add numerics dependency to prototype.

Commit:
Start building Tensor

Commit:
Add a dense tensor.

Commit:
WIP, accomodating a more dynamic tensor type.

Considering how to dispatch operations and accomodate laziness.

Commit:
Split things up and comment them.

Commit:
Experiment with static tensor element definition

Commit:
Checkpoint

Commit:
Checkpoint.

Commit:
Untwist the use of static function requirements on _AnyTensorElement.

The idea had been to keep requirements off the instances, but that doesn't work
with existentials, so dropping it for now.

Commit:
Comment TensorElement.swift

Commit:
Further cleanup.

Commit:
More testing

Commit:
Add subscripting ot DynamicTensor

Commit:
WIP on EagerDynamicTensor

Commit:
Fix assertion in to accomodate scalar TensorElements.

It's not entirely clear whether scalars should be subscriptable with
[0] rather than []... something to ponder.

Commit:
Add withUnsafe[Mutable]Pointer stuff to FixedSizeArray
